emulator/periph/src/flash_ctrl.rs

```diff
@@ -306,74 +299,78 @@ impl DummyFlashCtrl {
-        self.ctrl_regwen.reg.modify(CtrlRegwen::En::SET);
     }
 }
 
 impl FlashPeripheral for DummyFlashCtrl {
-    // Poll the flash controller to process the operation
     fn poll(&mut self) {
-        // Check if the operation is pending
-        if self.operation_pending {
+        if self.timer.fired(&mut self.operation_start) {
             self.process_io();
-            self.operation_pending = false;
         }
     }
 
     fn warm_reset(&mut self) {}
     fn update_reset(&mut self) {}
 
-    fn read_interrupt_state(
+    fn read_fl_interrupt_state(
         &mut self,
         _size: emulator_types::RvSize,
     ) -> emulator_bus::ReadWriteRegister<
         u32,
-        registers_generated::flash_ctrl::bits::InterruptState::Register,
+        registers_generated::flash_ctrl::bits::FlInterruptState::Register,
     > {
         emulator_bus::ReadWriteRegister::new(self.interrupt_state.reg.get())
     }
 
-    fn write_interrupt_state(
+    fn write_fl_interrupt_state(
         &mut self,
         _size: emulator_types::RvSize,
         _val: emulator_bus::ReadWriteRegister<
             u32,
-            registers_generated::flash_ctrl::bits::InterruptState::Register,
+            registers_generated::flash_ctrl::bits::FlInterruptState::Register,
         >,
     ) {
-        // Interrupt state register: write 1 to clear
-        if _val.reg.is_set(InterruptState::Error) {
-            self.interrupt_state
-                .reg
-                .modify(InterruptState::Error::CLEAR);
+        // Interrupt state register: SW write 1 to clear
+        if _val.reg.is_set(FlInterruptState::Error) {
+            self.clear_interrupt(FlashCtrlIntType::Error);
         }
-        if _val.reg.is_set(InterruptState::Event) {
-            self.interrupt_state
-                .reg
-                .modify(InterruptState::Event::CLEAR);
+        if _val.reg.is_set(FlInterruptState::Event) {
+            self.clear_interrupt(FlashCtrlIntType::Event);
         }
     }
 
-    fn read_interrupt_enable(
+    fn read_fl_interrupt_enable(
         &mut self,
         _size: emulator_types::RvSize,
     ) -> emulator_bus::ReadWriteRegister<
         u32,
-        registers_generated::flash_ctrl::bits::InterruptEnable::Register,
+        registers_generated::flash_ctrl::bits::FlInterruptEnable::Register,
     > {
         emulator_bus::ReadWriteRegister::new(self.interrupt_enable.reg.get())
     }
 
-    fn write_interrupt_enable(
+    fn write_fl_interrupt_enable(
         &mut self,
         _size: emulator_types::RvSize,
         _val: emulator_bus::ReadWriteRegister<
             u32,
-            registers_generated::flash_ctrl::bits::InterruptEnable::Register,
+            registers_generated::flash_ctrl::bits::FlInterruptEnable::Register,
         >,
     ) {
+        if self.interrupt_state.reg.is_set(FlInterruptState::Error)
+            && _val.reg.is_set(FlInterruptEnable::Error)
+        {
+            self._error_irq.set_level(true);
```

Comment:
You also want to check if the IRQ is current set, but the `val` is now `CLEAR`, you want to `set_level(false)` to stop raising the interrupt.